src/main_frontend.cpp

```diff
@@ -0,0 +1,614 @@
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdio.h>
+#include <string.h>
+#define SDL_MAIN_HANDLED
+#include "SDL.h"
+#include "mcu.h"
+#include "mcu_opcodes.h"
+#include "mcu_interrupt.h"
+#include "mcu_timer.h"
+#include "pcm.h"
+#include "lcd.h"
+#include "submcu.h"
+#include "midi.h"
+#include "utf8main.h"
+#include "utils/files.h"
+#include "emu.h"
+#include "ringbuffer.h"
+#include "math_util.h"
+
+#if __linux__
+#include <unistd.h>
+#include <limits.h>
+#endif
+
+struct fe_emu_instance_t {
+    emu_t        emu;
+    ringbuffer_t sample_buffer;
+    SDL_Thread*  thread;
+    bool         running;
+};
+
+const size_t FE_MAX_INSTANCES = 16;
+
+struct frontend_t {
+    fe_emu_instance_t instances[FE_MAX_INSTANCES];
+    size_t instances_in_use = 0;
+
+    int audio_buffer_size = 0;
+    int audio_page_size = 0;
+
+    SDL_AudioDeviceID sdl_audio = 0;
+};
+
+bool FE_AllocateInstance(frontend_t& container, fe_emu_instance_t** result)
+{
+    if (container.instances_in_use == FE_MAX_INSTANCES)
+    {
+        return false;
+    }
+
+    fe_emu_instance_t& fe = container.instances[container.instances_in_use];
+    memset(&fe, 0, sizeof(fe_emu_instance_t));
+    ++container.instances_in_use;
+
+    if (result)
+    {
+        *result = &fe;
+    }
+
+    return true;
+}
+
+void FE_SendMIDI(frontend_t& fe, size_t n, uint8_t* first, uint8_t* last)
+{
+    while (first != last)
+    {
+        MCU_PostUART(*fe.instances[n].emu.mcu, *first);
+        ++first;
+    }
+}
+
+void FE_BroadcastMIDI(frontend_t& fe, uint8_t* first, uint8_t* last)
+{
+    for (size_t i = 0; i < fe.instances_in_use; ++i)
+    {
+        FE_SendMIDI(fe, i, first, last);
+    }
+}
+
+void FE_RouteMIDI(frontend_t& fe, uint8_t* first, uint8_t* last)
+{
+    if (*first < 0x80)
+    {
+        printf("FE_RouteMIDI received data byte %02x\n", *first);
+        return;
+    }
+
+    const bool is_sysex = *first == 0xF0;
+    const uint8_t channel = *first & 0x0F;
+
+    if (is_sysex)
+    {
+        FE_BroadcastMIDI(fe, first, last);
+    }
+    else
+    {
+        FE_SendMIDI(fe, channel % fe.instances_in_use, first, last);
+    }
+}
+
+void FE_ReceiveSample(void* userdata, int *sample)
+{
+    fe_emu_instance_t& fe = *(fe_emu_instance_t*)userdata;
+    sample[0] >>= 15;
+    if (sample[0] > INT16_MAX)
+        sample[0] = INT16_MAX;
+    else if (sample[0] < INT16_MIN)
+        sample[0] = INT16_MIN;
+    sample[1] >>= 15;
+    if (sample[1] > INT16_MAX)
+        sample[1] = INT16_MAX;
+    else if (sample[1] < INT16_MIN)
+        sample[1] = INT16_MIN;
+    RB_Write(fe.sample_buffer, sample[0], sample[1]);
+}
```

Comment:
IMO, you're overthinking this a bit and micro-optimising early... You could just simple pass `AudioFrame` structs around (or similar). Doing something like this is perfectly fine, so just using simple values:

`AudioFrame RenderNextSample() { ... }`

The compiler is more than capable of getting rid unnecessary copies, plus you'll get guaranteed copy elision and return value optimisation unless you're using some archaic compiler.

https://en.wikipedia.org/wiki/Copy_elision

So no need to complicate life with pointer and such (which can make things run _slower_ in fact, because unsafe pointers can point to well anything, which makes the compilers job a lot harder than just expressing _intent_ by returning simple structs/values).

Reply:
Also, you could simply try it and re-run your nice test suite that you used for the measurements. My guess is you won't see any performance degradation.

Reply:
Nice improvement by introducing `audio_frame_t` 👍🏻